src/runner.py

```diff
@@ -0,0 +1,576 @@
+                "systemd-docker-proxy.j2"
+            ).render(proxies=self.config.proxies)
+
+            # Set docker daemon proxy config
+            docker_service_path = Path("/etc/systemd/system/docker.service.d")
+            docker_service_proxy = docker_service_path / "http-proxy.conf"
+
+            self.instance.files.mk_dir(str(docker_service_path))
+            self._put_file(str(docker_service_proxy), docker_proxy_contents)
+
+            self.instance.execute(["systemctl", "daemon-reload"])
+            self.instance.execute(["systemctl", "restart", "docker"])
+
+            # Set docker client proxy config
+            docker_client_proxy = {
+                "proxies": {
+                    "default": {
+                        "httpProxy": self.config.proxies["http"],
+                        "httpsProxy": self.config.proxies["https"],
+                        "noProxy": self.config.proxies["no_proxy"],
+                    }
+                }
+            }
+            docker_client_proxy_content = json.dumps(docker_client_proxy)
+            # Configure the docker client for root user and ubuntu user.
+            self._put_file("/root/.docker/config.json", docker_client_proxy_content)
+            self._put_file("/home/ubuntu/.docker/config.json", docker_client_proxy_content)
+            self.instance.execute(
+                ["/usr/bin/chown", "-R", "ubuntu:ubuntu", "/home/ubuntu/.docker"]
+            )
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def _register_runner(self, registration_token: str, labels: Sequence[str]) -> None:
+        """Register the runner on GitHub.
+
+        Args:
+            registration_token: Registration token request from GitHub.
+            labels: Labels to tag the runner with.
+        """
+        if self.instance is None:
+            raise RunnerError("Runner operation called prior to runner creation.")
+
+        logger.info("Registering runner %s", self.config.name)
+
+        register_cmd = [
+            "/usr/bin/sudo",
+            "-u",
+            "ubuntu",
+            str(self.config_script),
+            "--url",
+            f"https://github.com/{self.config.path.path()}",
+            "--token",
+            registration_token,
+            "--ephemeral",
+            "--unattended",
+            "--labels",
+            ",".join(labels),
+            "--name",
+            self.instance.name,
+        ]
+
+        if isinstance(self.config.path, GitHubOrg):
+            register_cmd += ["--runnergroup", self.config.path.group]
+
+        self.instance.execute(
+            register_cmd,
+            cwd=str(self.runner_application),
+        )
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def _start_runner(self) -> None:
+        """Start the GitHub runner."""
+        if self.instance is None:
+            raise RunnerError("Runner operation called prior to runner creation.")
```

Comment:
The complexity of the runner class comes from it is possible to have runner without a LXD instance and not registered on GitHub, runner with LXD instance that is registered on GitHub, runner without a LXD instance and registered to GitHub (the LXD instance might crashed), etc.

It is possible to represent all these states as different type of objects with [state pattern](https://en.wikipedia.org/wiki/State_pattern), which would avoid assert done in the line 496, 497. This would require a major rewrite so I did not did it.

Reply:
Major comments